core/pta/device.c

```diff
@@ -53,6 +60,23 @@ static TEE_Result get_devices(uint32_t types,
 		}
 	}
 
+#ifdef CFG_EARLY_TA
+	for_each_early_ta(eta) {
+			if (!(eta->flags & TA_FLAG_DEVICE_ENUM))
```

Comment:
It seems that this part can be moved into a helper function which can be used here and in the other loop above instead.

Reply:
yes, 2 loops on pseudo TAs and early are the same. I can combine this 2 functions.

Reply:
looked to this place. doing helper will bring more complexity to code read due to both cycles change ip_size, op_size and do memcpy. Also moving if into helper function is very strange idea for me. This macros is used only in 2 places it's better to not split it and use the same code.

Reply:
I'd prefer something along:
``` C
static void add_ta(uint32_t flags, const TEE_UUID *uuid, uint8_t *buf, 
                   size_t blen, size_t *pos)
{
        if (flags & TA_FLAG_DEVICE_ENUM) {
                if (*pos + sizeof(*uuid) < blen)
                        tee_uuid_to_octets(buf + *pos, uuid);
                (*pos) += sizeof(*uuid);
        }
}

...

        SCATTERED_ARRAY_FOREACH(ta, pseudo_tas, struct pseudo_ta_head)
                add_ta(ta->flags, &ta->uuid, buf, blen, &pos);

        if (IS_ENABLED(CFG_EARLY_TA))
                for_each_early_ta(eta)
                        add_ta(eta->flags, &eta->uuid, buf, blen, &pos);
...

        if (pos > blen)
                return TEE_ERROR_SHORT_BUFFER;

        return TEE_SUCCESS;

Reply:
Looks better and efficient.

Reply:
While writing in that way I see strange bug with 3.7 (which I can compile under OE and run.)

get_devices() is called onece with params[0].memref.size = 0 then it sets params[0].memref.size += sizeof(TEE_UUID); and returns TEE_ERROR_SHORT_BUFFER.

Then it called second time with params[0].memref.size  = 0x10. And then TA added. I'm trying to understand if that is expected.

Reply:
I have not checked the caller, but what you are describing is a common pattern when the caller does not know how many items to expect and therefore the size of the buffer to allocate. So, not a bug as far as I can tell ;-)

Reply:
Ok, will send updated version soon.